### PR TITLE
Allow shim's dirname in path and args

### PR DESCRIPTION
**Pull request by felipecrs**

### CODE REVIEW

**Comment by felipecrs** [comment]
I'm trying to fix https://github.com/71/scoop-better-shimexe/issues/10, but I have no clue on what I'm doing, I'd really appreciate a little help. :(

**Comment by felipecrs** [comment]
The current solution is actually working:

```sh-session
➜ cat C:\Users\felip\AppData\Local\Volta\tools\image\npm\7.9.0\bin\npm.shim
path = C:\Users\felip\AppData\Local\Volta\tools\image\node\14.16.1\node.EXE
args = %~dp0\npm-cli.js
➜ C:\Users\felip\AppData\Local\Volta\tools\image\npm\7.9.0\bin\npm.exe --version
7.9.0
```
Any tips for improving the performance?

**Comment by ImportTaste** [comment]
RegEx isn't really something you use when you're trying to get the fastest performance possible.

I don't know anything about C++, but maybe @kiennq could vouch for using a function like this? https://www.cplusplus.com/reference/string/string/replace/ It's just a simple replace operation, no RegEx required.

**Comment by kiennq** [comment]
Sorry for the late reply, and yes, this feature looks useful. It would be great if you can make it ready for merge :)

File: shim.cpp [hunk 1]
```diff
@@ -50,12 +51,20 @@ namespace std
     typedef optional<wstring> wstring_p;
 }
 
+std::wstring GetDirectory(wchar_t* exe) {
```
Comment (kiennq) [inline shim.cpp:54]: I would use `wstring_view` to minimize the creation of new strings as much as possible
```suggestion
std::wstring_view GetDirectory(std::wstring_view exe) {
```

File: shim.cpp [hunk 2]
```diff
@@ -50,12 +51,20 @@ namespace std
...
+    std::wstring exeString(exe);
+    std::wstring::size_type pos = std::wstring(exeString).find_last_of(L"\\/");
```
Comment (kiennq) [inline shim.cpp:56]:
```suggestion
    auto pos = std::wstring(exeString).find_last_of(L"\\/");
```

File: shim.cpp [hunk 3]
```diff
@@ -50,12 +51,20 @@ namespace std
...
+    return std::wstring(exeString).substr(0, pos);
```
Comment (kiennq) [inline shim.cpp:57]:
```suggestion
    return exeString.substr(0, pos);
```

File: shim.cpp [hunk 4]
```diff
@@ -105,7 +114,9 @@ std::tuple<std::wstring_p, std::wstring_p> GetShimInfo()
             continue;
         }
     }
-
+    std::wstring key = std::wstring(L"%~dp0");
+    path.emplace(std::regex_replace(path.value(), std::wregex(key), dirname));
```
Comment (kiennq) [inline shim.cpp:118]: Please use `replace` instead of `regex` version, which seems to be overkilled

Comment (kiennq) [inline shim.cpp:118]: Do the path needed to be replaced with the placeholder string? It seems to be overkilled

File: shim.cpp [hunk 5]
```diff
@@ -105,7 +114,9 @@ std::tuple<std::wstring_p, std::wstring_p> GetShimInfo()
             continue;
         }
     }
-
+    std::wstring key = std::wstring(L"%~dp0");
```
Comment (kiennq) [inline shim.cpp:117]: You can probably spin these 3 lines to a different function, for example
``` c++
std::tuple<std::wstring_p, std::wstring_p> NormalizeShimInfo(std::wstring_p& path, std::wstring_p& args)
{
    static constexpr auto s_dirPlaceHolder = L"%~dp0"sv;
    auto dirname = GetDirectory(path);
    args->replace(path->find(s_dirPlaceHolder), s_dirPlaceHolder.size(), dirname);
    return {path, args};
}
```

File: shim.cpp (hunk 4 above)
Comment (felipecrs) [reply shim.cpp:118]: Honestly, I don't remember why I did that.

**Comment by felipecrs** [comment]
@kiennq, first, thanks for the review. You did everything!

Being completely honest with you, I blindly applied all your suggestions, but I have no clue whether they work or not. I have almost zero knowledge about C/C++, and I'm struggling a lot to even being able to build it.

Would you mind taking the PR by yourself? I have given permission to you to edit it.

My goal with the PR was to just highlight the use-case as a PoC, and I'm happy it served that purpose.

File: shim.cpp (hunk 5 above)
Comment (ImportTaste) [reply shim.cpp:117]: @felipecrs I was going to compile and test out the PR on your behalf, but I noticed you didn't implement this suggestion. Please do, and remember to remove `import <regex>` from the top as well.

File: shim.cpp (hunk 4 above)
Comment (ImportTaste) [reply shim.cpp:118]: @felipecrs I was going to compile and test out the PR on your behalf, but I noticed you didn't implement this suggestion. Please do, and remember to remove `import <regex>` from the top as well.

Comment (felipecrs) [reply shim.cpp:118]: I tried, but I could not think of way to do. If you know how to do it, I would appreciate if you could do it.

Comment (felipecrs) [reply shim.cpp:118]: I invited you to my fork, so that you can change push to the PR if you need.

**Comment by ImportTaste** [comment]
@kiennq I would also like to voice support for integrating the code, I hit a situation today where relative pathing would really be useful.

**Comment by felipecrs** [comment]
Not from my side. :)

**Comment by felipecrs** [comment]
Thanks a ton, @kiennq!

**Comment by kiennq** [comment]
@ImportTaste It works for me though.
image

Are you using the latest pre-built one?
It doesn't contain this change yet. Let me kick off a new build.

**Pull request closed without merging**